clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -261,24 +266,25 @@ public boolean isTransactional() {
         return transactionalId != null;
     }
 
-    public synchronized boolean hasPartitionsToAdd() {
+    synchronized boolean hasPartitionsToAdd() {
         return !newPartitionsInTransaction.isEmpty() || !pendingPartitionsInTransaction.isEmpty();
     }
 
-    public synchronized boolean isCompletingTransaction() {
+    synchronized boolean isCompletingTransaction() {
         return currentState == State.COMMITTING_TRANSACTION || currentState == State.ABORTING_TRANSACTION;
     }
 
-    public synchronized boolean hasError() {
+    synchronized boolean hasError() {
         return currentState == State.ABORTABLE_ERROR || currentState == State.FATAL_ERROR;
     }
 
-    public synchronized boolean isAborting() {
+    synchronized boolean isAborting() {
         return currentState == State.ABORTING_TRANSACTION;
     }
 
-    synchronized boolean isInTransaction() {
-        return currentState == State.IN_TRANSACTION || isCompletingTransaction();
+    synchronized boolean hasOngoingTransaction() {
+        // transactions are considered ongoing once started until completion or a fatal error
+        return currentState == State.IN_TRANSACTION || isCompletingTransaction() || hasAbortableError();
```

Comment:
Actually we can simplify this since, for the time being, transactional producers can only send data inside of a transaction. However, I think we should keep `hasAbortableError`. We consider the transaction ongoing as long as the client still has the capability to complete it.